### PR TITLE
Add `Vec::pop_chunk` (as unstable)

### PR DESCRIPTION
**Pull request by scottmcm**
This isn't a particularly complicated bit of unsafe code, but being able to get multiple items in vec order with something like
```rust
let [a, b, c] = v.pop_chunk()?;
```
is much nicer than doing a dance with multiple `pop`s.

(Especially since `[v.pop()?, v.pop()?]` gives the opposite order and can partially remove things.)

Tracking issue #95217

This does not include a `push_chunk`, since now that arrays are `IntoIterator`, that can be done with `.extend([a, b, c])`.

I chose to call this `pop_chunk` instead of `split_off_chunk` because `split_off(k)` is an index, not a length.

### CODE REVIEW

**Comment by leonardo-m** [comment]
I appreciate your work here, but what can an use case be? :)

File: library/alloc/src/vec/mod.rs [hunk 1]
```diff
@@ -1763,6 +1763,42 @@ impl<T, A: Allocator> Vec<T, A> {
...
+    /// assert_eq!(vec.pop_chunk::<2>(), None); // Not enough elements
+    /// assert_eq!(vec, [1]); // So the vector was unchanged
+    ///
+    /// let mut vec: Vec<i32> = Vec::new();
+    /// assert_eq!(vec.pop_chunk::<0>(), Some([]));
+    /// ```
+    #[inline]
+    #[unstable(feature = "vec_pop_chunk", issue = "95217")]
+    pub fn pop_chunk<const N: usize>(&mut self) -> Option<[T; N]> {
+        let new_len = self.len.checked_sub(N)?;
```
Comment (scottmcm) [inline library/alloc/src/vec/mod.rs:1791]: Turns out the `checked_sub` saves an instruction compared to the ordinary way: <https://rust.godbolt.org/z/Pvrdffvjz>

Comment (the8472) [reply library/alloc/src/vec/mod.rs:1791]: It doesn't seem to vectorize well. https://rust.godbolt.org/z/o3vzjcGrE

But for loops an owning `IntoArrayIter` for `Vec` might be a better choice anyway, we currently only have the borrowing ones via the slice deref.

**Comment by Mark-Simulacrum** [comment]
The implementation here seems OK, so r=me in that respect.

However, I am a little cautious with such an API -- it seems to me that a more general approach here would be drain(..) with a next_chunk and next_rchunk method on the Drain iterator that does this from either end.

Are you expecting to add similar methods to VecDeque on either end?

Those are mostly minor nits than strong concerns though.

**Pull request closed without merging**